File: practice2/typescript/src/domain/ADTDirectedGraph.ts
export type Vertex = number;
export type EdgeCost = number;
export type Edge = [Vertex, Vertex];

export type Vertices = Vertex[];
export type Outbound = Map<Vertex, Vertices>;
export type Inbound = Map<Vertex, Vertices>;
export type EdgeCosts = Map<string, EdgeCost>;

export type OutboundIterator = Iterator<Vertex>;
export type EdgesIterator = Iterator<[Edge, EdgeCost]>;
export type VerticesIterator = Iterator<Vertex>;

import { ADTDirectedGraphIterator } from "./ADTDirectedGraphIterator";

export class ADTDirectedGraph {
  private outbound: Outbound;
  private inbound: Inbound;
  private costs: EdgeCosts;

  private edgeToKey(edge: Edge): string {
    return `${edge[0]},${edge[1]}`;
  }

  private keyToEdge(key: string): Edge {
    const parts = key.split(",").map(Number);
    return [parts[0], parts[1]];
  }

  /**
   * @brief Gets the keys of the outbound map. This should only be used by the iterator.
   *
   * @returns MapIterator<Vertex>
   */
  public getOutboundKeys(): IterableIterator<Vertex> {
    return this.outbound.keys();
  }

  /**
   * @brief Construct a new ADTDirectedGraph object.
   */
  constructor() {
    this.outbound = new Map<Vertex, Vertices>();
    this.inbound = new Map<Vertex, Vertices>();
    this.costs = new Map<string, EdgeCost>();
  }

  /**
   * @brief Returns the number of vertices in the graph.
   *
   * @return number
   */
  nbVertices(): number {
    return this.outbound.size;
  }

  /**
   * @brief Returns the number of edges in the graph.
   *
   * @return number
   */
  nbEdges(): number {
    return this.costs.size;
  }

  /**
   * @brief Checks if an edge between 2 vertexes exist.
   *
   * @param first First vertex.
   * @param second Second vertex.
   * @return true There is an edge between the vertexes.
   * @return false There is no edge between the vertexes.
   */
  isEdge(first: Vertex, second: Vertex): boolean {
    if (!this.outbound.has(first) || !this.inbound.has(second)) {
      return false;
    }

    const outboundNeigh = this.outbound.get(first)!;
    return outboundNeigh.includes(second);
  }

  /**
   * @brief Get the in-degree of a vertex.
   *
   * @param vertex The vertex in question.
   * @return number
   */
  getInDegree(vertex: Vertex): number {
    return this.inbound.get(vertex)?.length ?? 0;
  }

  /**
   * @brief Get the out-degree of a vertex.
   *
   * @param vertex The vertex in question.
   * @return number
   */
  getOutDegree(vertex: Vertex): number {
    return this.outbound.get(vertex)?.length ?? 0;
  }

  /**
   * @brief Adds an edge to the graph.
   *
   * @param edge Edge to be added.
   * @param cost Cost associated with the edge.
   * @return true If add was successful.
   * @return false If add was not successful.
   */
  addEdge(edge: Edge, cost: EdgeCost): boolean {
    // check if edge vertices exist in the graph
    if (!this.outbound.has(edge[0]) || !this.outbound.has(edge[1])) {
      return false;
    }

    const edgeKey = this.edgeToKey(edge);

    // check if edge is already in the graph
    if (this.costs.has(edgeKey)) {
      return false;
    }

    // handle outbound
    this.outbound.get(edge[0])!.push(edge[1]);
    // handle inbound
    this.inbound.get(edge[1])!.push(edge[0]);
    // handle the cost map
    this.costs.set(edgeKey, cost);

    return true;
  }

  /**
   * @brief Removes an edge from the graph.
   *
   * @param edge The edge to be removed.
   * @return true If edge was removed successfully.
   * @return false If edge was not removed successfully.
   */
  removeEdge(edge: Edge): boolean {
    const edgeKey = this.edgeToKey(edge);

    // check if edge exists in the graph
    if (!this.costs.has(edgeKey)) {
      return false;
    }

    // handle outbound
    const outboundNeigh = this.outbound.get(edge[0])!;
    const outIndex = outboundNeigh.indexOf(edge[1]);

    if (outIndex !== -1) {
      outboundNeigh.splice(outIndex, 1);
    }

    // handle inbound
    const inboundNeigh = this.inbound.get(edge[1])!;
    const inIndex = inboundNeigh.indexOf(edge[0]);

    if (inIndex !== -1) {
      inboundNeigh.splice(inIndex, 1);
    }

    // handle removing the edge from the costs
    this.costs.delete(edgeKey);

    return true;
  }

  /**
   * @brief Adds a vertex to the graph.
   *
   * @param vertex The vertex to be added.
   * @return true If adding the vertex was successful.
   * @return false If adding the vertex was not successful.
   */
  addVertex(vertex: Vertex): boolean {
    // check if vertex is in graph already
    if (this.outbound.has(vertex)) {
      return false;
    }

    // handle outbound
    this.outbound.set(vertex, []);
    // handle inbound
    this.inbound.set(vertex, []);

    return true;
  }

  /**
   * @brief Removes a vertex from the graph.
   *
   * @param vertex The vertex to be removed.
   * @return true If removing the vertex was successful.
   * @return false If removing the vertex was not successful.
   */
  removeVertex(vertex: Vertex): boolean {
    // check if vertex is in the graph
    if (!this.outbound.has(vertex)) {
      return false;
    }

    // handle inbound first
    const outboundNeigh = this.outbound.get(vertex)!;
    for (const currOutVertex of outboundNeigh) {
      const currInboundNeigh = this.inbound.get(currOutVertex)!;
      const index = currInboundNeigh.indexOf(vertex);
      if (index !== -1) {
        currInboundNeigh.splice(index, 1);
      }
    }

    // delete the key from outbound
    this.outbound.delete(vertex);

    // handle outbound
    const inboundNeigh = this.inbound.get(vertex)!;
    for (const currInVertex of inboundNeigh) {
      const currOutboundNeigh = this.outbound.get(currInVertex)!;
      const index = currOutboundNeigh.indexOf(vertex);
      if (index !== -1) {
        currOutboundNeigh.splice(index, 1);
      }
    }

    // delete the key from inbound
    this.inbound.delete(vertex);

    // handle costs - remove alll
    const keysToDelete: string[] = [];
    for (const [edgeKey, _] of this.costs) {
      const edge = this.keyToEdge(edgeKey);
      if (edge[0] === vertex || edge[1] === vertex) {
        keysToDelete.push(edgeKey);
      }
    }
    for (const key of keysToDelete) {
      this.costs.delete(key);
    }

    return true;
  }

  /**
   * @brief Returns the iterator for the graph's vertices.
   *
   * @return ADTDirectedGraphIterator
   */
  parseVertices(): ADTDirectedGraphIterator {
    return new ADTDirectedGraphIterator(this);
  }

  /**
   * @brief Returns the iterator for the outbound neighbors of a given vertex from the beginning.
   *
   * @throws Throws an error if given vertex is not in the graph.
   * @param vertex The vertex in question.
   * @return VerticesIterator
   */
  parseOutboundOfGivenVertexBegin(vertex: Vertex): VerticesIterator {
    if (!this.outbound.has(vertex)) {
      throw new Error("Vertex not found in graph");
    }

    const vertices = this.outbound.get(vertex)!;
    return vertices[Symbol.iterator]();
  }

  /**
   * @brief Returns the iterator for the outbound neighbors of a given vertex from the end.
   *
   * @throws Throws an error if given vertex is not in the graph.
   * @param vertex The vertex in question.
   * @return VerticesIterator
   */
  parseOutboundOfGivenVertexEnd(vertex: Vertex): VerticesIterator {
    if (!this.outbound.has(vertex)) {
      throw new Error("Vertex not found in graph");
    }

    const exhaustedIterator: VerticesIterator = {
      next(): IteratorResult<Vertex> {
        return { done: true, value: undefined };
      },
    };
    return exhaustedIterator;
  }

  /**
   * @brief Returns the iterator for the inbound neighbors of a given vertex from the beginning.
   *
   * @throws Throws an error if given vertex is not in the graph.
   * @param vertex The vertex in question.
   * @return VerticesIterator
   */
  parseInboundOfGivenVertexBegin(vertex: Vertex): VerticesIterator {
    if (!this.inbound.has(vertex)) {
      throw new Error("Vertex not found in graph");
    }

    const vertices = this.inbound.get(vertex)!;
    return vertices[Symbol.iterator]();
  }

  /**
   * @brief Returns the iterator for the inbound neighbors of a given vertex from the end.
   *
   * @throws Throws an error if given vertex is not in the graph.
   * @param vertex The vertex in question.
   * @return VerticesIterator
   */
  parseInboundOfGivenVertexEnd(vertex: Vertex): VerticesIterator {
    if (!this.inbound.has(vertex)) {
      throw new Error("Vertex not found in graph");
    }

    const exhaustedIterator: VerticesIterator = {
      next(): IteratorResult<Vertex> {
        return { done: true, value: undefined };
      },
    };
    return exhaustedIterator;
  }

  /**
   * @brief Returns the iterator for the edges (with costs) from the beginning.
   *
   * @return EdgesIterator
   */
  parseEdgesBegin(): EdgesIterator {
    const entries = Array.from(this.costs.entries());
    let index = 0;

    const graphSelf = this;

    return {
      next(): IteratorResult<[Edge, EdgeCost]> {
        if (index < entries.length) {
          const [key, cost] = entries[index++];
          const edge = graphSelf.keyToEdge(key);
          return { done: false, value: [edge, cost] };
        }
        return { done: true, value: undefined };
      },
    };
  }

  /**
   * @brief Returns the iterator for the edges (with costs) from the end.
   *
   * @return EdgesIterator
   */
  parseEdgesEnd(): EdgesIterator {
    return {
      next(): IteratorResult<[Edge, EdgeCost]> {
        return { done: true, value: undefined };
      },
    };
  }
}

File: practice2/cpp/src/domain/ADTDirectedGraph.cpp
#include "ADTDirectedGraph.h"
#include "ADTDirectedGraphIterator.h"

#include <iostream>
#include <set>
#include <utility>
#include <queue>
#include <algorithm>
#include <fstream>
#include <random>
#include <climits>
#include <vector>
#include <exception>

ADTDirectedGraph::ADTDirectedGraph()
{
    // handle outbound
    this->outbound = std::map<Vertex, std::vector<Vertex>>{};

    // handle inbound
    this->inbound = std::map<Vertex, std::vector<Vertex>>{};

    // handle costs
    this->costs = std::map<Edge, EdgeCost>{};
}

unsigned int ADTDirectedGraph::nbVertices() const
{
    unsigned int nbVertices = 0;

    for (const auto &_ : this->outbound)
    {
        nbVertices++;
    }

    return nbVertices;
}

unsigned int ADTDirectedGraph::nbEdges() const
{
    return this->costs.size();
}

bool ADTDirectedGraph::isEdge(Vertex first, Vertex second) const
{

    if (this->outbound.count(first) == 0 || this->inbound.count(second) == 0)
    {
        return false;
    }

    std::vector<Vertex> outboundNeigh = this->outbound.at(first);

    bool found = false;

    for (const Vertex &vertex : outboundNeigh)
    {
        if (vertex == second)
        {
            found = true;
            break;
        }
    }

    return found;
}

unsigned int ADTDirectedGraph::getInDegree(Vertex vertex) const
{
    return this->inbound.at(vertex).size();
}

unsigned int ADTDirectedGraph::getOutDegree(Vertex vertex) const
{
    return this->outbound.at(vertex).size();
}

bool ADTDirectedGraph::addEdge(Edge edge, EdgeCost cost)
{

    // check if edge vertices exist in the graph
    if (this->outbound.count(edge.first) == 0 || this->outbound.count(edge.second) == 0)
    {
        return false;
    }

    // check if edge is already in the graph
    // we can more easily do this by checking the costs map
    if (this->costs.count(edge) == 1)
    {
        return false;
    }

    // handle outbound
    this->outbound.at(edge.first).push_back(edge.second);
    // handle inbound
    this->inbound.at(edge.second).push_back(edge.first);
    // handle the cost map
    this->costs[edge] = cost;

    return true;
}

bool ADTDirectedGraph::removeEdge(Edge edge)
{

    // check if edge exists in the graph
    // easiest way to do this is through the costs
    if (this->costs.count(edge) == 0)
    {
        return false;
    }

    // handle outbound
    unsigned int outIndex = -1;
    std::vector<Vertex> &outboundNeigh = this->outbound.at(edge.first);
    for (unsigned int i = 0; i < outboundNeigh.size(); i++)
    {
        if (outboundNeigh.at(i) == edge.second)
        {
            outIndex = i;
            break;
        }
    }
    // remove the vertex from the outbound neighbors of first
    outboundNeigh.erase(outboundNeigh.begin() + outIndex);

    // handle inbound
    // basically the same thing as above but we change the source and dest
    unsigned int inIndex = -1;
    std::vector<Vertex> &inboundNeigh = this->inbound.at(edge.second);
    for (unsigned int i = 0; i < inboundNeigh.size(); i++)
    {
        if (inboundNeigh.at(i) == edge.first)
        {
            inIndex = i;
            break;
        }
    }
    // remove the vertex from the inbound neighbors of second
    inboundNeigh.erase(inboundNeigh.begin() + inIndex);

    // handle removing the edge from the costs
    this->costs.erase(edge);

    return true;
}

bool ADTDirectedGraph::addVertex(Vertex vertex)
{
    // check if vertex is in graph already
    if (this->outbound.count(vertex) != 0)
    {
        return false;
    }

    // handle outbound
    this->outbound[vertex] = std::vector<Vertex>{};

    // handle inbound
    this->inbound[vertex] = std::vector<Vertex>{};

    return true;
}

bool ADTDirectedGraph::removeVertex(Vertex vertex)
{

    // check if vertex is in the graphs
    if (this->outbound.count(vertex) == 0)
    {
        return false;
    }

    // handle inbound first
    std::vector<Vertex> &outboundNeigh = this->outbound.at(vertex);

    // we check in the inbound of the current vertex if we have any matches then delete said match
    for (const Vertex &currOutVertex : outboundNeigh)
    {
        std::vector<Vertex> &currInboundNeigh = this->inbound.at(currOutVertex);

        for (unsigned int i = 0; i < currInboundNeigh.size(); i++)
        {
            if (currInboundNeigh.at(i) == vertex)
            {
                currInboundNeigh.erase(currInboundNeigh.begin() + i);
                break;
            }
        }
    }

    // handle inbound
    std::vector<Vertex> &inboundNeigh = this->inbound.at(vertex);

    // we check in the outbound of the current vertex if we have any matches then delete said match
    for (const Vertex &currInVertex : inboundNeigh)
    {
        std::vector<Vertex> &currOutboundNeigh = this->outbound.at(currInVertex);

        for (unsigned int i = 0; i < currOutboundNeigh.size(); i++)
        {
            if (currOutboundNeigh.at(i) == vertex)
            {
                currOutboundNeigh.erase(currOutboundNeigh.begin() + i);
                break;
            }
        }
    }

    // now we delete the key from the inbound itself
    this->inbound.erase(vertex);

    // now we delete the key from the outbound itself
    this->outbound.erase(vertex);

    // handle costs

    // we iterate over the costs then if the key (type Edge) includes match we delete the entry
    // compared to practice 1 this time we actually handle deleting properly when using iterators
    for (auto it = this->costs.begin(); it != this->costs.end();)
    {
        if (it->first.first == vertex || it->first.second == vertex)
        {
            it = this->costs.erase(it);
        }
        else
        {
            ++it;
        }
    }

    return true;
}

ADTDirectedGraphIterator ADTDirectedGraph::parseVertices() const
{

    return ADTDirectedGraphIterator(*this);
}

std::vector<Vertex>::const_iterator ADTDirectedGraph::parseOutboundOfGivenVertexBegin(Vertex vertex) const
{

    if (this->outbound.count(vertex) == 0)
    {
        throw std::exception();
    }

    return this->outbound.at(vertex).begin();
}

std::vector<Vertex>::const_iterator ADTDirectedGraph::parseOutboundOfGivenVertexEnd(Vertex vertex) const
{

    if (this->outbound.count(vertex) == 0)
    {
        throw std::exception();
    }

    return this->outbound.at(vertex).end();
}

std::vector<Vertex>::const_iterator ADTDirectedGraph::parseInboundOfGivenVertexEnd(Vertex vertex) const
{

    if (this->inbound.count(vertex) == 0)
    {
        throw std::exception();
    }

    return this->inbound.at(vertex).end();
}

std::vector<Vertex>::const_iterator ADTDirectedGraph::parseInboundOfGivenVertexBegin(Vertex vertex) const
{

    if (this->inbound.count(vertex) == 0)
    {
        throw std::exception();
    }

    return this->inbound.at(vertex).begin();
}

std::map<Edge, EdgeCost>::const_iterator ADTDirectedGraph::parseEdgesBegin() const
{
    return this->costs.begin();
}

std::map<Edge, EdgeCost>::const_iterator ADTDirectedGraph::parseEdgesEnd() const
{
    return this->costs.end();
}

std::pair<std::set<Vertex>, std::map<Vertex, Vertex>> ADTDirectedGraph::forwardBreadthFirstTraversal(ADTDirectedGraph graph, Vertex start)
{
    std::queue<Vertex> queue;
    std::map<Vertex, Vertex> prev;
    std::map<Vertex, Vertex> dist;
    std::set<Vertex> visited;

    queue.push(start);
    visited.insert(start);
    dist[start] = 0;
    prev[start] = start;

    while (!queue.empty())
    {
        // those 2 lines are equivalent to the q.enqueue operation
        Vertex x = queue.front();
        queue.pop();

        std::vector<Vertex>::const_iterator neighOutBegin = graph.parseOutboundOfGivenVertexBegin(x);
        std::vector<Vertex>::const_iterator neighOutEnd = graph.parseOutboundOfGivenVertexEnd(x);

        while (neighOutBegin != neighOutEnd)
        {
            Vertex y = *neighOutBegin;

            if (visited.find(y) == visited.end())
            {
                queue.push(y);
                visited.insert(y);
                dist[y] = dist[x] + 1;
                prev[y] = x;
            }

            ++neighOutBegin;
        }
    }

    return {visited, prev};
}

std::vector<Vertex> ADTDirectedGraph::findLowestLengthPathBetweenTwoVertices(ADTDirectedGraph graph, Vertex start, Vertex end)
{
    auto [visited, prev] = forwardBreadthFirstTraversal(graph, start);

    // we couldnt reach the end
    if (visited.find(end) == visited.end())
    {
        return {};
    }

    std::vector<Vertex> path;
    Vertex current = end;

    while (current != start)
    {
        path.push_back(current);
        current = prev[current];
    }
    path.push_back(start);

    // we used std::vector for the path meaning we can apply this alg
    std::reverse(path.begin(), path.end());

    return path;
}

bool ADTDirectedGraph::readGraph(ADTDirectedGraph &graph, const std::string &filename)
{
    std::ifstream inputFile(filename);

    // forgot to reset the damn graph
    graph = ADTDirectedGraph{};

    if (!inputFile.is_open())
    {
        return false;
    }

    unsigned int nbVertices, nbEdges;
    inputFile >> nbVertices >> nbEdges;

    Vertex first, second;
    EdgeCost cost;

    for (unsigned int i = 0; i < nbEdges; i++)
    {
        inputFile >> first >> second >> cost;

        graph.addVertex(first);
        graph.addVertex(second);

        graph.addEdge({first, second}, cost);
    }

    return true;
}

bool ADTDirectedGraph::writeGraph(const ADTDirectedGraph &graph, const std::string &filename)
{
    std::ofstream outputFile(filename);

    if (!outputFile.is_open())
    {
        return false;
    }

    outputFile << graph.nbVertices() << ' ' << graph.nbEdges() << std::endl;

    std::map<Edge, EdgeCost>::const_iterator startIt = graph.parseEdgesBegin();
    std::map<Edge, EdgeCost>::const_iterator endIt = graph.parseEdgesEnd();

    while (startIt != endIt)
    {

        Edge currEdge = startIt->first;
        EdgeCost currCost = startIt->second;

        outputFile << currEdge.first << ' ' << currEdge.second << ' ' << currCost << std::endl;

        ++startIt;
    }

    return true;
}

ADTDirectedGraph ADTDirectedGraph::generateRandomGraph(unsigned int nbVertices, unsigned int nbEdges)
{

    // so if we have nbEdges > nbVertices * (nbVertices - 1) we have a problemo
    if (nbEdges > nbVertices * (nbVertices - 1))
    {
        throw std::exception();
    }

    ADTDirectedGraph graph = ADTDirectedGraph{};

    // way of generating numbers from 0 to the limit of unsigned int
    // rd is a seed gen, gen is a mersenne twister engine
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<unsigned int> randomUnsignedInt(0, UINT_MAX);
    std::uniform_int_distribution<unsigned int> randomVerticesIndex(0, nbVertices - 1);

    std::vector<Vertex> generatedVertices = std::vector<Vertex>{};

    // we generate nbVertices number of vertices
    for (unsigned int i = 0; i < nbVertices; ++i)
    {
        unsigned int vertex = Vertex{randomUnsignedInt(gen)};
        graph.addVertex(vertex);
        generatedVertices.push_back(vertex);
    }

    // we handle generating random edges
    unsigned int generatedEdges = 0;
    while (generatedEdges < nbEdges)
    {

        Vertex firstVertex = generatedVertices.at(randomVerticesIndex(gen));
        Vertex secondVertex = generatedVertices.at(randomVerticesIndex(gen));

        // if we got unlucky and we already have an edge we just skip
        if (graph.isEdge(firstVertex, secondVertex))
        {
            continue;
        }

        // we create an edge with unsigned int cost as well
        unsigned int randomCost = EdgeCost{randomUnsignedInt(gen)};
        Edge createdEdge = Edge{firstVertex, secondVertex};
        graph.addEdge(createdEdge, randomCost);

        generatedEdges++;
    }

    return graph;
}
